Source/Core/Core/Debugger/CodePatch.h

```diff
@@ -0,0 +1,44 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#pragma once
+
+#include <vector>
+
+#include "Common/CommonTypes.h"
+
+class CodePatchBase
+{
+public:
+  enum class State
+  {
+    SOURCE,
+    PATCH,
+    INVALID
+  };
+  virtual ~CodePatchBase();
+  virtual u32 GetAddress() const = 0;
+  virtual u32 GetSize() const = 0;
+  virtual bool Apply() = 0;
+  virtual bool Remove() = 0;
+  virtual State GetState() const = 0;
+};
+
+class CodePatch : public CodePatchBase
+{
+public:
+  CodePatch(u32 address, const std::vector<u8>& patch);
```

Comment:
This can also have a constructor of the form

```cpp
CodePatch(u32 address, std::vector<u8>&& patch)
```

which moves the vector.

This should also have a defaulted destructor in the cpp file to prevent inlining of teardown code for the std::vector members.